soup.go

```diff
@@ -15,20 +15,67 @@ import (
 	"golang.org/x/net/html"
 )
 
-// Root is a structure containing a pointer to an html node, the node value, and an error variable to return an error if occurred
+// ErrorType defines types of errors that are possible from soup
+type ErrorType int
+
+const (
+	// ErrUnableToParse will be returned when the HTML could not be parsed
+	ErrUnableToParse ErrorType = iota
+	// ErrElementNotFound will be returned when element was not found
+	ErrElementNotFound
+	// ErrNoNextSibling will be returned when no next sibling can be found
+	ErrNoNextSibling
+	// ErrNoPreviousSibling will be returned when no previous sibling can be found
+	ErrNoPreviousSibling
+	// ErrNoNextElementSibling will be returned when no next element sibling can be found
+	ErrNoNextElementSibling
+	// ErrNoPreviousElementSibling will be returned when no previous element sibling can be found
+	ErrNoPreviousElementSibling
+	// ErrCreatingGetRequest will be returned when the get request couldn't be created
+	ErrCreatingGetRequest
+	// ErrInGetRequest will be returned when there was an error during the get request
+	ErrInGetRequest
+	// ErrReadingResponse will be returned if there was an error reading the response to our get request
+	ErrReadingResponse
+)
+
+// Error allows easier introspection on the type of error returned.
+// If you know you have a Error, you can compare the Type to one of the exported types
+// from this package to see what kind of error it is, then further inspect the Error() method
+// to see if it has more specific details for you, like in the case of a ErrElementNotFound
+// type of error.
+type Error struct {
+	Type ErrorType
+	msg  string
```

Comment:
Since golang doesn't support inheritance but allows composition, i thought we could have it like this
```go
type Error struct {
    Type ErrorType
    error
}
```
and access it like `Error.Type` and `Error.Error()` (the public method exposed by the golang interface). But i just tried this on golang playground, and it looks like we can compose our struct with other __structs__, but not __interfaces__ (`error` is an interface) :/

Reply:
IMO the way it actually works is even better than this, as what makes something an error is just that it has a method with a signature like `Error() string`. It's even more powerful than embedding an error 👍